zebra-state/src/service/check/utxo.rs

```diff
@@ -0,0 +1,108 @@
+    finalized_state: &FinalizedState,
+) -> Result<(), ValidateContextError> {
+    let mut block_spends = HashSet::new();
+
+    for (spend_tx_index_in_block, transaction) in prepared.block.transactions.iter().enumerate() {
+        let spends = transaction.inputs().iter().filter_map(|input| match input {
+            transparent::Input::PrevOut { outpoint, .. } => Some(outpoint),
+            // Coinbase inputs represent new coins,
+            // so there are no UTXOs to mark as spent.
+            transparent::Input::Coinbase { .. } => None,
+        });
+
+        for spend in spends {
+            // check for in-block duplicate spends
+            if block_spends.contains(spend) {
+                return Err(DuplicateTransparentSpend {
+                    out_point: *spend,
+                    location: "the same block",
+                });
+            }
+
+            // check spends occur in chain order
+            //
+            // because we are in the non-finalized state, we need to check spends within the same block,
+            // the unspent and spent non-finalized UTXOs, and the finalized UTXOs.
+
+            if let Some(output) = prepared.new_outputs.get(spend) {
+                // reject the spend if it uses an output from this block,
+                // but the output was not created by an earlier transaction
+                //
+                // we know the spend is invalid, because transaction IDs are unique
+                //
+                // transaction IDs also commit to transaction inputs,
+                // so it should be cryptographically impossible for a transaction
+                // to spend its own outputs
+                if output.tx_index_in_block >= spend_tx_index_in_block {
+                    return Err(EarlyTransparentSpend { out_point: *spend });
+                }
+            } else {
+                // reject the spend if its UTXO is not available in the block,
+                // non-finalized parent chain, or finalized state
+                if !parent_chain_unspent_utxos.contains_key(spend)
+                    && finalized_state.utxo(spend).is_none()
+                {
+                    if parent_chain_spent_utxos.contains(spend) {
+                        return Err(DuplicateTransparentSpend {
+                            out_point: *spend,
+                            location: "the parent non-finalized chain",
+                        });
+                    } else {
+                        // we don't keep spent UTXOs for the finalized state,
+                        // so all we can say is that it's missing
+                        // (from both the finalized and non-finalized chains)
+                        return Err(MissingTransparentOutput {
+                            out_point: *spend,
+                            location: "the parent chain",
+                        });
+                    }
+                }
+            }
```

Comment:
Nit: I was thinking if it would be possible to simplify the if-else logic, but I'm not sure if I'm missing something :thinking: The following seems equivalent, and it removes the need for the `parent_chain_unspent_utxos` parameter.

```suggestion
            if let Some(output) = prepared.new_outputs.get(spend) {
                // reject the spend if it uses an output from this block,
                // but the output was not created by an earlier transaction
                //
                // we know the spend is invalid, because transaction IDs are unique
                //
                // transaction IDs also commit to transaction inputs,
                // so it should be cryptographically impossible for a transaction
                // to spend its own outputs
                if output.tx_index_in_block >= spend_tx_index_in_block {
                    return Err(EarlyTransparentSpend { out_point: *spend });
                }
            } else if parent_chain_spent_utxos.contains(spend) {
                // reject the spend if its UTXO is already spent in the
                // non-finalized parent chain
                return Err(DuplicateTransparentSpend {
                    out_point: *spend,
                    location: "the parent non-finalized chain",
                });
            } else if finalized_state.utxo(spend).is_none() {
                // we don't keep spent UTXOs for the finalized state,
                // so all we can say is that it's missing
                // (from both the finalized and non-finalized chains)
                return Err(MissingTransparentOutput {
                    out_point: *spend,
                    location: "the parent chain",
                });
            }
```

Reply:
I think we can simplify this code now we know that UTXOs are unique in a single chain.

Also I think there might be a missing "else" case (no UTXO in the non-finalized state), but I'll add it after I merge your suggestion.

Here's some background:

I wrote the first draft of this code when we were unsure about transaction ID uniqueness. So I was trying to allow for a UTXO to be spent, then added again. (But the original code is buggy anyway.)

But then the spec got clarified:
https://github.com/zcash/zips/commit/7a8b12d945e73abede4203823ffd1687c7fb5f3d

Reply:
See 4979f193f9790e3f5022062b5c3535283da1f897 and 76fbce1e36b7cc9e86aee2000aee8dc6a39fa5cc.

Here are the changes I made:
- add a missing check against the non-finalized unspent UTXOs
- split each error case into a separate check
- combine `contains` and `insert`
- rename arguments and edit error strings for clarity

Thanks for the feedback, the code is a lot clearer now.